File: server/controllers/paymentController.js
const { query } = require('../config/db');
const { processDemoPayment } = require('../services/paymentService');

const processPayment = async (req, res, next) => {
  try {
    const {
      bookingId,
      amount,
      paymentMethod = 'UPI (Simulated)',
      paymentType = 'ADVANCE',
      notes
    } = req.body;

    if (!bookingId || !amount) {
      return res.status(400).json({
        success: false,
        message: 'Booking ID and payment amount are required.'
      });
    }

    const [bookings] = await query('SELECT * FROM bookings WHERE id = ?', [Number(bookingId)]);
    if (!bookings || bookings.length === 0) {
      return res.status(404).json({ success: false, message: 'Booking not found.' });
    }

    const paymentResult = await processDemoPayment({
      bookingId: Number(bookingId),
      amount: Number(amount),
      paymentMethod,
      paymentType,
      notes: notes || `Simulated ${paymentType} payment via ${paymentMethod}`
    });

    res.status(200).json({
      success: true,
      message: 'Payment simulation successful. Booking confirmed!',
      payment: paymentResult
    });
  } catch (err) {
    next(err);
  }
};

const getPaymentByBookingId = async (req, res, next) => {
  try {
    const bookingId = Number(req.params.bookingId);
    const [payments] = await query('SELECT * FROM payments WHERE booking_id = ? ORDER BY payment_date DESC', [bookingId]);

    res.json({
      success: true,
      count: payments.length,
      payments: payments || []
    });
  } catch (err) {
    next(err);
  }
};

module.exports = {
  processPayment,
  getPaymentByBookingId
};

File: server/controllers/payment_controller.py
from flask import request
from config.db_config import SessionLocal
from models.models import Booking, Payment
from services.payment_service import process_demo_payment
from utils.helpers import success_response, error_response

def handle_demo_payment(current_user):
    data = request.get_json() or {}
    booking_id = data.get('booking_id')
    payment_type = data.get('payment_type', 'ADVANCE')
    payment_method = data.get('payment_method', 'UPI (Simulated)')

    if not booking_id:
        return error_response("Booking ID is required", 400)

    db = SessionLocal()
    try:
        booking = db.query(Booking).filter(Booking.id == booking_id).first()
        if not booking:
            return error_response("Booking not found", 404)

        if booking.user_id != current_user.id and current_user.role != 'ADMIN':
            return error_response("Unauthorized", 403)

        if booking.booking_status == 'CANCELLED':
            return error_response("Cannot process payment for a cancelled booking", 400)

        result = process_demo_payment(
            db=db,
            booking=booking,
            payment_type=payment_type,
            payment_method=payment_method
        )
        return success_response(result, "Simulated payment successful! Reference generated.", 200)

    except Exception as e:
        db.rollback()
        return error_response(f"Payment processing failed: {str(e)}", 500)
    finally:
        db.close()

def get_payments_for_booking(booking_id, current_user):
    db = SessionLocal()
    try:
        booking = db.query(Booking).filter(Booking.id == booking_id).first()
        if not booking:
            return error_response("Booking not found", 404)

        payments = db.query(Payment).filter(Payment.booking_id == booking_id).order_by(Payment.payment_date.desc()).all()
        return success_response({
            'booking_reference': booking.booking_reference,
            'total_amount': float(booking.total_amount),
            'advance_amount': float(booking.advance_amount),
            'payment_status': booking.payment_status,
            'payments': [p.to_dict() for p in payments]
        }, "Payment records retrieved", 200)
    finally:
        db.close()
